hooks/checkout

```diff
@@ -7,30 +7,37 @@ DIR="$(cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd)"
 # shellcheck source=lib/plugin.bash
 . "$DIR/../lib/plugin.bash"
 
-CHECKOUT_PATHS=$(plugin_read_config PATHS "")
-
-if [ -z "${CHECKOUT_PATHS}" ]; then
-  echo 'Missing `paths` option in the plugin'
-  exit 1
+if plugin_read_list_into_result PATHS; then
+    CHECKOUT_PATHS=("${result[@]}")
+else
+    echo "Missing 'paths' option in the plugin"
+    exit 1
 fi
 
-echo "Creating sparse-checkout with paths: ${CHECKOUT_PATHS}"
+echo "Scanning SSH keys for remote git repository"
+[[ -d ~/.ssh ]] || mkdir -p ~/.ssh
+ssh-keyscan "${BUILDKITE_REPO_SSH_HOST}" >> ~/.ssh/known_hosts
+
+echo "Creating sparse-checkout with paths: ${CHECKOUT_PATHS[*]}"
 
 # clone the repo without checking out files if it does not exist already
-if [ ! -d .git ]; then
-    git clone -v --depth 1 --no-checkout ${BUILDKITE_REPO} .
+if [[ ! -d .git ]]; then
+    git clone \
+        --depth 1 \
+        --filter=blob:none \
+        --no-checkout \
+        ${BUILDKITE_REPO_MIRROR:+--reference "$BUILDKITE_REPO_MIRROR"} \
+        -v \
+        "${BUILDKITE_REPO}" .
 fi
 
 git clean -ffxdq
 # fetch branch if commit is HEAD
-if [ "${BUILDKITE_COMMIT}" = "HEAD" ]; then
-    git fetch origin ${BUILDKITE_BRANCH}
-    git checkout FETCH_HEAD
+if [[ ${BUILDKITE_COMMIT} = "HEAD" ]]; then
+    git fetch --depth 1 origin "${BUILDKITE_BRANCH}"
 else
-    git fetch origin ${BUILDKITE_COMMIT}
-    git checkout ${BUILDKITE_COMMIT}
```

Comment:
This makes sense to remove in a way, but its actually still causing the wrong build message in the UI. I think there is some other flow we need to do but I haven't fully researched or tried it out yet.
Basically, without this, the hook is not actually checking out the commit for the build given by `BUILDKITE_COMMIT`. So then below when you fetch, add sparse checkout and then checkout again, all its doing is creating a sparse checkout at the master/main branch because it hasnt switched to the given commit yet.

Reply:
Perhaps using `git switch` is intended for this. Or perhaps even just leave the checkout in place and it won't touch files since its cloned without a checkout. We'll do some testing on our side

Reply:
If this is needed then the final `git checkout` on the last line of this script maybe should become:
```
if [[ ${BUILDKITE_COMMIT} = "HEAD" ]]; then
    git checkout FETCH_HEAD
else
    git checkout "${BUILDKITE_COMMIT}"
fi
```

Reply:
Yeah I think either way should work. Whether leaving the checkouts in the conditionals here, or reproducing it below after the sparse checkout config

Reply:
Can't leave the checkouts here, or else it'll check all files out since the sparse-checkout stuff hasn't been set in motion yet.